File: 49. Group Anagrams/solution.js
/**
 * @param {string[]} strs
 * @return {string[][]}
 */
var groupAnagrams = function (strs) {
  const groups = {};

  for (s of strs) {
    const count = Array(26).fill(0);
    for (c of s) {
      count[c.charCodeAt(0) - "a".charCodeAt(0)]++;
    }
    if (groups[count] === undefined) {
      groups[count] = [];
    }
    groups[count].push(s);
  }
  return Object.values(groups);
};

// Time complexity: O(n * m)
// Space complexity: O(n * m)

File: 49. Group Anagrams/Solution.cs
public class Solution {
    public IList<IList<string>> GroupAnagrams(string[] strs) {
        IDictionary<string, IList<string>> groups = new Dictionary<string, IList<string>>();

        foreach (String s in strs) {
            int[] count = new int[26];
            foreach (char c in s.ToCharArray()) {
                count[c - 'a']++;
            }
            string key = string.Join(",", count);
            if (!groups.ContainsKey(key)) {
                groups[key] = new List<string>();
            }
            groups[key].Add(s);
        }
        return new List<IList<string>>(groups.Values);
    }
}

// Time complexity: O(N * K)
// Space complexity: O(N * K)